runtime/contract_update_validation.go

```diff
@@ -0,0 +1,456 @@
+		nonStorableType: refType,
+		Range:           ast.NewRangeFromPositioned(refType),
+	}
+}
+
+func (validator *ContractUpdateValidator) checkNameEquality(expectedType *ast.NominalType, foundType *ast.NominalType) bool {
+	isExpectedQualifiedName := expectedType.IsQualifiedName()
+	isFoundQualifiedName := foundType.IsQualifiedName()
+
+	// A field with a composite type can be defined in two ways:
+	// 	- Using type name (var x @ResourceName)
+	//	- Using qualified type name (var x @ContractName.ResourceName)
+
+	if isExpectedQualifiedName && !isFoundQualifiedName {
+		return validator.checkIdentifierEquality(expectedType, foundType)
+	}
+
+	if isFoundQualifiedName && !isExpectedQualifiedName {
+		return validator.checkIdentifierEquality(foundType, expectedType)
+	}
+
+	// At this point, either both are qualified names, or both are simple names.
+	// Thus, do a one-to-one match.
+	if expectedType.Identifier.Identifier != foundType.Identifier.Identifier {
+		return false
+	}
+
+	return identifiersEqual(expectedType.NestedIdentifiers, foundType.NestedIdentifiers)
+}
+
+func (validator *ContractUpdateValidator) checkIdentifierEquality(
+	qualifiedNominalType *ast.NominalType,
+	simpleNominalType *ast.NominalType,
+) bool {
+
+	// Situation:
+	// qualifiedNominalType -> identifier: A, nestedIdentifiers: [foo, bar, ...]
+	// simpleNominalType -> identifier: foo,  nestedIdentifiers: [bar, ...]
+
+	// If the first identifier (i.e: 'A') refers to a composite decl that is not the enclosing contract,
+	// then it must be referring to an imported contract. That means the two types are no longer the same.
+	if qualifiedNominalType.Identifier.Identifier != validator.newDeclaration.DeclarationIdentifier().Identifier {
+		return false
+	}
+
+	if qualifiedNominalType.NestedIdentifiers[0].Identifier != simpleNominalType.Identifier.Identifier {
+		return false
+	}
+
+	return identifiersEqual(simpleNominalType.NestedIdentifiers, qualifiedNominalType.NestedIdentifiers[1:])
+}
+
+func (validator *ContractUpdateValidator) checkConformances(
+	oldEnum *ast.CompositeDeclaration,
+	newEnum *ast.CompositeDeclaration,
+) error {
+
+	oldConformances := oldEnum.Conformances
+	newConformances := newEnum.Conformances
+
+	if len(oldConformances) != len(newConformances) {
+		return &ConformanceCountMismatchError{
+			expected: len(oldConformances),
+			found:    len(newConformances),
+			Range:    ast.NewRangeFromPositioned(newEnum.Identifier),
+		}
+	}
+
+	for index, conformance := range oldConformances {
+		err := conformance.CheckEqual(newConformances[index], validator)
+		if err != nil {
+			return &ConformanceMismatchError{
+				err:   err,
+				Range: ast.NewRangeFromPositioned(conformance),
+			}
+		}
+	}
+
+	return nil
+}
+
+func getTypeMismatchError(expectedType ast.Type, foundType ast.Type) *TypeMismatchError {
+	return &TypeMismatchError{
+		expectedType: expectedType,
+		foundType:    foundType,
+		Range:        ast.NewRangeFromPositioned(expectedType),
```

Comment:
The error position should be for the new type, not the old one:

```suggestion
		Range:        ast.NewRangeFromPositioned(foundType),
```